api/client/batch_client.py

```diff
@@ -30,61 +55,81 @@ def __init__(self, api_host, access_token):
-                            params,
-                            json.loads(e.response.body.decode("utf-8"))['data'][0])
-                        self._logger.warning(
-                            'Redirecting {} to {}'.format(params, new_params),
-                            extra=log_record)
+                try:
+                    response = yield self._http_client.fetch(http_request)
+                    status_code = response.code
+                except HTTPError as e:
+                    # Catch non-200 codes that aren't errors
+                    status_code = e.code if hasattr(e, 'code') else None
+                    if status_code in [204, 206]:
+                        log_msg = {204: 'No Content', 206: 'Partial Content'}[status_code]
+                        response = e.response
+                        log_request(start_time, retry_count, log_msg, status_code)
+                        # Do not retry.
+                    elif status_code == 301:
+                        redirected_ids = json.loads(e.response.body.decode("utf-8"))['data'][0]
+                        new_params = lib.redirect(params, redirected_ids)
+                        log_request(start_time, retry_count,
+                                    'Redirecting {} to {}'.format(params, new_params), status_code)
                         params = new_params
-                else:
-                    self._logger.error(e.response.error, extra=log_record)
-                    raise Exception('Giving up on {} after {} tries. \
-                        Error is: {}.'.format(
-                        url, retry_count, e.response.error))
-
+                        continue  # retry
+                    else:  # Otherwise, propagate to error handling
+                        raise e
+            except Exception as e:
+                # HTTPError raised when there's a non-200 status code
+                # socket.gaio error raised when there's a connection error
+                response = e.response if hasattr(e, 'response') else e
+                status_code = e.code if hasattr(e, 'code') else None
+                error_msg = e.response.error if (hasattr(e, 'response') and
+                                                 hasattr(e.response, 'error')) else e
+                log_request(start_time, retry_count, error_msg, status_code)
+                if status_code in [429, 500, 503, 504]:
+                    # First retry is immediate.
+                    # After that, exponential backoff before retrying.
+                    if retry_count > 1:
+                        time.sleep(2 ** retry_count)
+                    continue
+                elif status_code in [400, 401, 402, 404]:
+                    break  # Do not retry. Go right to raising an Exception.
+
+            # Request was successful
+            log_request(start_time, retry_count, 'OK', status_code)
+            raise gen.Return(json_decode(response.body) if hasattr(response, 'body') else None)
```

Comment:
why json_decode?

Reply:
`json_decode` was always being called, but I moved it up from `get_data_points()` to here since the `raise gen.Return(data.body)` was breaking on 204 No Content responses since they don't have a body.

lib.py equivalent is `resp.json()` here https://github.com/gro-intelligence/api-client/blob/fc39206bf5c81fe63721d9bcc834899add6824a0/api/client/lib.py#L555 but tornado's body is bytes instead of a string, so it doesn't have that method. Have to decode and parse json, like `json.loads(response.body.decode("utf-8"))` to convert it to a python object. Tornado provides `json_decode` that just does those two things in one function.